pkg/cmd/auth/shared/git_credential.go

```diff
@@ -1,47 +1,63 @@
 package shared
 
 import (
-	"bytes"
-	"context"
 	"errors"
-	"fmt"
-	"path/filepath"
-	"strings"
 
-	"github.com/MakeNowJust/heredoc"
 	"github.com/cli/cli/v2/git"
-	"github.com/cli/cli/v2/internal/ghinstance"
-	"github.com/google/shlex"
+	"github.com/cli/cli/v2/pkg/cmd/auth/shared/gitcredentials"
 )
 
 type GitCredentialFlow struct {
-	Executable string
-	Prompter   Prompt
-	GitClient  *git.Client
+	Prompter Prompt
+
+	HelperConfig *gitcredentials.HelperConfig
+	Updater      *gitcredentials.Updater
 
 	shouldSetup bool
-	helper      string
+	helper      gitcredentials.Helper
 	scopes      []string
 }
 
 func (flow *GitCredentialFlow) Prompt(hostname string) error {
-	var gitErr error
-	flow.helper, gitErr = gitCredentialHelper(flow.GitClient, hostname)
-	if isOurCredentialHelper(flow.helper) {
+	// First we'll fetch the credential helper that would be used for this host
+	var configuredHelperErr error
+	flow.helper, configuredHelperErr = flow.HelperConfig.ConfiguredHelper(hostname)
+	// If the helper is gh itself, then we don't need to ask the user if they want to update their git credentials
+	// because it will happen automatically by virtue of the fact that gh will return the active token.
+	//
+	// Since gh is the helper, this token may be used for git operations, so we'll additionally request the workflow
+	// scope to ensure that git push operations that include workflow changes succeed.
+	if flow.helper.IsOurs() {
 		flow.scopes = append(flow.scopes, "workflow")
 		return nil
 	}
 
+	// Prompt the user for whether they want to configure git with the newly obtained token
 	result, err := flow.Prompter.Confirm("Authenticate Git with your GitHub credentials?", true)
 	if err != nil {
 		return err
 	}
 	flow.shouldSetup = result
 
 	if flow.shouldSetup {
-		if isGitMissing(gitErr) {
-			return gitErr
+		// If the user does want to configure git, we'll check the error returned from fetching the configured helper
+		// above. If the error indicates that git isn't installed, we'll return an error now to ensure that the auth
+		// flow is aborted before the user goes any further.
+		//
+		// Note that this is _slightly_ naive because there may be other reasons that fetching the configured helper
+		// fails that might cause later failures but this code has existed for a long time and I don't want to change
+		// it as part of a refactoring.
+		//
+		// Refs:
+		//  * https://git-scm.com/docs/git-config#_description
+		//  * https://github.com/cli/cli/pull/4109
+		var errNotInstalled *git.NotInstalled
+		if errors.As(err, &errNotInstalled) {
+			return configuredHelperErr
 		}
+
```

Comment:
**How can we reach this code block if the previous conditional is returning `err`?**

https://github.com/cli/cli/blob/d75548a63016b9001215a6e223cd82a1273e26d5/pkg/cmd/auth/shared/git_credential.go#L35-L39

Should this be `configuredHelperErr`?

```suggestion
		var errNotInstalled *git.NotInstalled
		if errors.As(configuredHelperErr, &errNotInstalled) {
			return configuredHelperErr
		}

```

Reply:
That's exactly right, great catch and was previously totally untested code, hence no test failures.

Reply:
In the original PR when I said:

> My larger goal here is to test these in a meaningfully integrated manner, and then call sites can uses fakes in their place.

This is somewhere that we would be able to stub that.